### PR TITLE
[llvm] Compile the kernel arguments to a StructType

### PR DESCRIPTION
**Pull request by lin-hitonami**
This PR does things as follows:
1. Add `char *arg_buffer` to `RuntimeContext` as the argument buffer.
2. Manage the memory of `arg_buffer` in `LaunchContextBuilder`.
3. Record `arg_buffer_size` and `args_type` to RuntimeContext for processing the arguments right before the kernel executes in `ModuleToFunctionConverter`. (They will be moved to `LaunchContextBuilder` as soon as the argument of the function returned by `ModuleToFunctionConverter` becomes `LaunchContextBuilder`.)
4. Deal with the argument compiled to struct type in LLVM in the codegen.
5. Allocate the argument buffer on device right before the kernel executes in `ModuleToFunctionConverter`.
6. Support f16

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #7576
* #7550
* __->__ #7441

### CODE REVIEW

**Comment by feisuzhu** [comment]
test

**Comment by feisuzhu** [comment]
ghstack bot failed: Check-run `Build and Test Windows`'s status `failure` is not success!

**Comment by feisuzhu** [comment]
ghstack bot failed: PR #7441 PR #7441 check-run `Build and Test Windows`'s status `failure` is not success!

**Comment by feisuzhu** [comment]
ghstack bot failed: No reviews found!

**Comment by feisuzhu** [comment]
ghstack bot failed: No reviews found!

**Comment by feisuzhu** [comment]
ghstack bot failed: PR #7441 is not approved yet!

**Review by feisuzhu — APPROVED** [review]
TESTING! NOT REALLY APPROVED!

**Comment by feisuzhu** [comment]
ghstack bot failed: PR #7441 check-run `Build and Test Windows`'s status `failure` is not success!

**Review by feisuzhu — CHANGES_REQUESTED** [review]
TESTING!

**Comment by feisuzhu** [comment]
ghstack bot failed: @feisuzhu has `CHANGES_REQUESTED` PR #7441, please resolve it first!

**Comment by feisuzhu** [comment]
ghstack bot failed: @feisuzhu has stamped PR #7441 `CHANGES_REQUESTED`, please resolve it first!

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 1]
```diff
@@ -2783,6 +2788,48 @@ void TaskCodeGenLLVM::set_struct_to_buffer(
                        current_element, current_index);
 }
 
+llvm::Value *TaskCodeGenLLVM::get_struct_arg(std::vector<int> index) {
+  auto *args_ptr = get_args_ptr(current_callable, get_context());
+  auto *args_type = current_callable->args_type;
+  auto *arg_type = args_type->get_element_type(index);
+  std::vector<llvm::Value *> gep_index;
```
Comment (bobcao3) [inline taichi/codegen/llvm/codegen_llvm.cpp:2795]: nit:
```suggestion
  std::vector<llvm::Value *> gep_index(index.size() + 1);
```
Then index into it

File: taichi/codegen/llvm/codegen_llvm.cpp [hunk 2]
```diff
@@ -2783,6 +2788,48 @@ void TaskCodeGenLLVM::set_struct_to_buffer(
...
+}
+void TaskCodeGenLLVM::set_args_ptr(Callable *callable,
+                                   llvm::Value *context,
+                                   llvm::Value *ptr) {
+  auto *runtime_context_type = get_runtime_type("RuntimeContext");
+  auto *args_type = tlctx->get_data_type(callable->args_type);
+  auto *zero = tlctx->get_constant(0);
+  auto *args_ptr =
+      builder->CreateGEP(runtime_context_type, context, {zero, zero});
+  args_ptr = builder->CreatePointerCast(
```
Comment (bobcao3) [inline taichi/codegen/llvm/codegen_llvm.cpp:2827]: ditto

File: taichi/util/bit.h [hunk 3]
```diff
@@ -47,6 +47,21 @@ TI_FORCE_INLINE float32 half_to_float(const uint16 x) {
                              0x007FE000)));  // sign : normalized : denormalized
 }
 
+TI_FORCE_INLINE uint16 float_to_half(const float32 x) {
```
Comment (bobcao3) [inline taichi/util/bit.h:50]: We have a F16 library submodule in the externals folder. Use that pls for IEEE correctness and hardware intrinsics on ARM CPUs

**Review by ailzhang — APPROVED** [review]
LGTM!

**Pull request closed without merging**